### ISSUE TITLE
Anubis - Potential Griefing Vulnerability in Smart Contract Deposit Cap Enforcement

### ISSUE THREAD

**Issue by sherlock-admin3** [sherlock-audit/2024-02-perpetual-judging#6]
Anubis

high

# Potential Griefing Vulnerability in Smart Contract Deposit Cap Enforcement

## Summary
The smart contract's **_checkDepositCap()** function contains a potential vulnerability where it only checks the contract's current balance against the deposit cap after a deposit has been made. This approach fails to preemptively block a deposit that would exceed the cap, instead only reverting if the cap is already breached. Consequently, an attacker might exploit the timing of their deposit to intentionally exceed the deposit cap, causing other legitimate transactions to fail with a **DepositCapExceeded** error. 
## Vulnerability Detail
The root cause of the potential griefing vulnerability in the smart contract deposit cap enforcement lies in the fact that the check for the deposit cap is done after the deposit has already been made. In the code snippet provided, the function _checkDepositCap() is called internally to check if the deposit cap has been exceeded after a deposit is made.

This allows an attacker to make a deposit that exceeds the cap, causing the contract to revert after the deposit has already been made. This can lead to a griefing attack where an attacker can intentionally cause the contract to revert, disrupting the normal operation of the contract and potentially causing loss of funds for legitimate users.

The vulnerability in the code lies in the fact that the deposit cap is checked after the deposit has already been made. This allows an attacker to deposit an amount exceeding the cap, causing the contract to revert after the deposit has been made.

**Proof of Concept (PoC) code:**

1. Deploy a smart contract with the following code:

```solidity
pragma solidity ^0.8.0;

contract VulnerableContract {
    uint256 public depositCap = 100;
    address public collateralToken;

    function deposit(uint256 amount) external {
        require(amount > 0, "Amount must be greater than 0");
        require(IERC20Metadata(collateralToken).balanceOf(address(this)) + amount <= depositCap, "Deposit cap exceeded");
        
        // Deposit logic
    }
}
```
2. Call the **deposit** function with an amount greater than the deposit cap (e.g., 150).
3. The deposit will be accepted and the contract will revert after the deposit has been made, allowing the attacker to bypass the deposit cap enforcement and potentially grief the contract by causing it to revert.

## Impact
 This vulnerability could be leveraged to perform a griefing attack, resulting in denial of service for honest users attempting to deposit funds into the contract.
## Code Snippet
https://github.com/sherlock-audit/2024-02-perpetual/blob/main/perp-contract-v3/src/vault/Vault.sol#L480-L485
## Tool used

Manual Review

## Recommendation
To fix this issue, the deposit cap check should be done before the deposit is made. This way, the contract can reject the deposit if it exceeds the cap before any funds are transferred.

Here is an example of how the code can be patched to fix the vulnerability:

```solidity
480       function _checkDepositCap(uint256 amount) internal view {
481           uint256 depositCap = _getConfig().getDepositCap();
482           if (IERC20Metadata(_getVaultStorage().collateralToken).balanceOf(address(this)) + amount > depositCap) {
483               revert LibError.DepositCapExceeded();
484           }
485       }
```
In this patched code, the **_checkDepositCap** function now takes an additional parameter **amount**, which represents the amount being deposited. The function checks if the total balance of the contract (including the new deposit amount) exceeds the deposit cap before allowing the deposit to proceed. This way, the deposit cap is enforced before the deposit is made, preventing any potential griefing attacks.

Duplicate of #10
